Aplib.Core/ExposedQueue.cs

```diff
@@ -0,0 +1,193 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Aplib.Core
+{
+    /// <summary>
+    /// A queue with all elements exposed.
+    /// Functionally works like a queue with indexing.
+    /// </summary>
+    public class ExposedQueue<T> : ICollection<T>
+    {
+        /// <summary>
+        /// The length of the array.
+        /// </summary>
+        public int MaxCount { get; private set; }
+
+        /// <summary>
+        /// Actual number of elements in the array.
+        /// Limited to <see cref="MaxCount"/>.
+        /// </summary>
+        public int Count
+        {
+            get => _count;
+            private set => _count = value > MaxCount ? MaxCount : value;
+        }
+
+        /// <inheritdoc/>
+        public bool IsReadOnly => false;
+
+        private int _count;
+
+        private readonly T[] _array;
+        private int _head;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExposedQueue{T}"/> class.
+        /// </summary>
+        /// <param name="size">The size of the array.</param>
+        public ExposedQueue(int size)
+        {
+            MaxCount = size;
+            _array = new T[MaxCount];
+            _head = MaxCount - 1;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExposedQueue{T}"/> class.
+        /// By default, assumes the array is filled.
+        /// </summary>
+        /// <param name="array">An array to use as the circular array.</param>
+        /// <param name="count">The number of actual elements in the array.</param>
+        public ExposedQueue(T[] array, int count = -1)
+        {
+            MaxCount = array.Length;
+            _array = array;
+            _head = MaxCount - 1;
+            Count = count == -1 ? MaxCount : count;
+        }
```

Comment:
Multiple pointers:
* The user could provide a negative count, which will not be caught in an exception.
* I would propose using nullables here. The default count of `-1` is not ideal. The user could always provide a count of -1 themselves. This should throw an exception instead of setting the count to the max count. The ideal solution would be to use `MaxCount` as default value, but since it is computed at runtime, this is not possible.

```suggestion
        public ExposedQueue(T[] array, int? count = null)
        {
            MaxCount = array.Length;
            _array = array;
            _head = MaxCount - 1;
            Count = count ?? MaxCount;
        }
```